sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java

```diff
@@ -0,0 +1,121 @@
+
+    private void cacheResponse(P page, Response response) {
+        if (page.getPageId() == null) {
+            return;
+        }
+        pages.put(page.getPageId(), new PageAndResponse<P>(page, response));
+
+        // setting previous page id should simplify implementation for androidx.arch.DataSource
+        final Iterator<PageAndResponse<P>> iterator = pages.values().iterator();
+        while(iterator.hasNext()){
+            final P existingPage = iterator.next().page;
+            if (page.getPageId().equals(existingPage.getNextPageId())){
+                page.setPreviousPageId(existingPage.getPageId());
+                break;
+            }
+        }
+    }
+
+    /**
+     * Gets first page in the collection along with the response retrieving the first page
+     * @param callback a callback interface for handling the first page and its response
+     */
+    public void getFirstPage(Callback<P> callback) {
+        if (firstPageId == null){
+            pagedDataRetriever.getFirstPage(new Callback<P>() {
+
+                @Override
+                public void onSuccess(P value, Response response) {
+                    if (value == null) {
+                        callback.onSuccess(value, response);
+                        return;
+                    }
+
+                    firstPageId = value.getPageId();
+                    cacheResponse(value, response);
+                    callback.onSuccess(value, response);
+                }
+
+                @Override
+                public void onFailure(Throwable t, Response response) {
+                    callback.onFailure(t, response);
+                }
+            });
+        }
+        else {
+            PageAndResponse<P> firstPageResponse = pages.get(firstPageId);
+            callback.onSuccess(firstPageResponse.page, firstPageResponse.response);
+        }
+    }
+
+    /**
+     * Retrieve a page by its id
+     * @param pageId id of the page
+     * @param callback callback interface for handling the page along with its response
+     */
+    public void getPage(String pageId, Callback<P> callback) {
+        PageAndResponse<P> pageEntry = pages.get(pageId);
+        if (pageEntry == null){
+            pagedDataRetriever.getPage(pageId, new Callback<P>() {
+                @Override
+                public void onSuccess(P value, Response response) {
+                    if (value == null) {
+                        callback.onSuccess(null, response);
+                    }
+
+                    cacheResponse(value, response);
+                    callback.onSuccess(value, response);
+                }
+
+                @Override
+                public void onFailure(Throwable t, Response response) {
+                    callback.onFailure(t, response);
+                }
+            });
+        }
+        else {
+            callback.onSuccess(pageEntry.page, pageEntry.response);
+        }
+    }
+
+    private static class PageAndResponse<P> {
```

Comment:
We should be using `com.azure.android.core.http.Response<T>` as the type to represent a response with an associated strongly-typed response body object. Is there any reason we aren't doing that here?

Reply:
Sorry Brandon, This branch is messed up somehow. Please put comment in https://github.com/Azure/azure-sdk-for-android/pull/317 instead. I'll still read here but more people may have been directed there.

Reply:
My bad. I thought this was the old one. ;-)

The reason for PageAndResponse is this collection has an interface where getFirstPage and getPage both take a Callback<P> where both Page and okhttp3.Response are supposed to be used. Unless we change the callback signature, we'd have to keep both in order to hand back the okhttp3.Response.

Reply:
Leaving as-is for this preview.

Reply:
Adding to this for later discussion: We can definitely handle the given `okhttp3.Response` in the callback and map its contents to our own `Response<T>` type. I think we did that in Storage at some point.